system/core/HyphaRequest.php

```diff
@@ -0,0 +1,131 @@
+	const HYPHA_SYSTEM_PAGE_INDEX = 'index';
+	const HYPHA_SYSTEM_PAGE_SETTINGS = 'settings';
+	const HYPHA_SYSTEM_PAGE_UPLOAD = 'upload';
+	const HYPHA_SYSTEM_PAGE_CHOOSER = 'chooser';
+
+	/** @var string */
+	private $requestQuery;
+
+	/** @var array */
+	private $isoLangList;
+
+	/** @var array */
+	private $requestParts;
+
+	/** @var null|false|string */
+	private $language;
+
+	/** @var null|false|string */
+	private $systemPage;
+
+	/**
+	 * @param $requestQuery
+	 * @param array $isoLangList Associative array containing connecting iso639 language codes with their native name (and english translation)
+	 */
+	public function __construct($requestQuery, array $isoLangList) {
+		$this->requestQuery = $requestQuery;
+		$this->isoLangList = $isoLangList;
+	}
+
+	public static function getSystemPages() {
+		return [
+			HyphaRequest::HYPHA_SYSTEM_PAGE_FILES,
+			HyphaRequest::HYPHA_SYSTEM_PAGE_IMAGES,
+			HyphaRequest::HYPHA_SYSTEM_PAGE_INDEX,
+			HyphaRequest::HYPHA_SYSTEM_PAGE_SETTINGS,
+			HyphaRequest::HYPHA_SYSTEM_PAGE_UPLOAD,
+			HyphaRequest::HYPHA_SYSTEM_PAGE_CHOOSER,
+		];
+	}
+
+	/**
+	 * @param bool $excludeLanguage
+	 *
+	 * @return string
+	 */
+	public function getRequestQuery($excludeLanguage = true) {
+		return implode('/', $this->getRequestParts($excludeLanguage));
+	}
+
+	/**
+	 * @param bool $excludeLanguage
+	 *
+	 * @return array
+	 */
+	public function getRequestParts($excludeLanguage = true) {
+		if ($this->requestParts === null) {
+			$this->requestParts = array_filter(explode('/', $this->requestQuery));
+		}
+
+		$requestParts = $this->requestParts;
+
+		if ($excludeLanguage) {
+			$language = $this->getLanguage();
+			if ($requestParts[0] === $language) {
+				array_shift($requestParts);
+			}
+		}
+
+		return $requestParts;
+	}
+
+	/**
+	 * Get the language from the request.
+	 *
+	 * @return null|string
+	 */
+	public function getLanguage() {
+		if ($this->language === null) {
+			$requestParts = $this->getRequestParts(false);
+
+			$firstPart = reset($requestParts);
+			if ($firstPart !== false && array_key_exists($firstPart, $this->isoLangList)) {
+				$this->language = $firstPart;
+			} else {
+				$this->language = false;
+			}
+		}
+
+		return $this->language === false ? null : $this->language;
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function isSystemPage() {
+		return $this->getSystemPage() !== null;
+	}
+
+	/**
+	 * @return string|null
+	 */
+	public function getSystemPage() {
+		if ($this->systemPage === null) {
+			$requestParts = $this->getRequestParts(false);
+			$systemPage = reset($requestParts);
+			if ($systemPage === false || !in_array($systemPage, HyphaRequest::getSystemPages())) {
+				$this->systemPage = false;
+			} else {
+				$this->systemPage = $systemPage;
+			}
+		}
+
+		return $this->systemPage === false ? null : $this->systemPage;
+	}
+
+	public function getPageType() {
+		if ($this->isSystemPage() || $this->getLanguage() !== null) {
+			return null;
```

Comment:
Doesn't an url with a language have a page type?

Reply:
Good catch, fixed it.

Reply:
... sorry.. the implementation was correct.. you cannot detect the page type from the query if there is a language specified.